app/uk/gov/hmrc/cgtpropertydisposalsfrontend/models/bpr/UnsuccessfulNameMatchAttempt.scala

```diff
@@ -16,19 +16,41 @@
 
 package uk.gov.hmrc.cgtpropertydisposalsfrontend.models.bpr
 
-import play.api.libs.json.{Json, OFormat}
-import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.ids.SAUTR
-import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.name.IndividualName
+import cats.Eq
+import julienrf.json.derived
+import play.api.libs.json.{Json, OFormat, Reads, Writes}
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.bpr.UnsuccessfulNameMatchAttempts.NameMatchDetails
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.ids.{SAUTR, TRN}
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.name.{IndividualName, TrustName}
 
-final case class UnsuccessfulNameMatchAttempts(
-  unsuccesfulAttempts: Int,
+final case class UnsuccessfulNameMatchAttempts[+A <: NameMatchDetails](
+  unsuccessfulAttempts: Int,
   maximumAttempts: Int,
-  lastNameTried: IndividualName,
-  lastSAUTRTried: SAUTR
+  lastDetailsTried: A
 )
 
 object UnsuccessfulNameMatchAttempts {
 
-  implicit val format: OFormat[UnsuccessfulNameMatchAttempts] = Json.format
+  sealed trait NameMatchDetails extends Product with Serializable
+
```

Comment:
I can see why you have gone down this approach but I think it is more complex than it needs to be for what it does. Essentially we just want to read the try-count from the cache for an individual or a trust. It seems that this type-level machinery could have been replaced by a simple method taking an enum to represent the entity type. Am I missing something?